server/asset/eth/coiner.go

```diff
@@ -0,0 +1,284 @@
+// available at least once before the swap be trusted for swap initializations.
+func newSwapCoin(backend *Backend, coinID []byte, sct swapCoinType) (*swapCoin, error) {
+	switch sct {
+	case sctInit, sctRedeem:
+	default:
+		return nil, fmt.Errorf("unknown swapCoin type: %d", sct)
+	}
+
+	cID, err := DecodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+	txCoinID, ok := cID.(*TxCoinID)
+	if !ok {
+		return nil, errors.New("coin ID not a txid")
+	}
+	txid := txCoinID.TxID
+	tx, _, err := backend.node.transaction(backend.rpcCtx, txid)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch transaction: %v", err)
+	}
+
+	txdata := tx.Data()
+	// Transactions that call contract functions must have extra data to do
+	// so.
+	if len(txdata) == 0 {
+		return nil, errors.New("tx calling contract functions should have extra data")
+	}
+
+	var (
+		counterParty       = new(common.Address)
+		secret, secretHash [32]byte
+		locktime           int64
+	)
+
+	switch sct {
+	case sctInit:
+		counterParty, secretHash, locktime, err = dexeth.ParseInitiateData(txdata)
+	case sctRedeem:
+		secret, secretHash, err = dexeth.ParseRedeemData(txdata)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse call data: %v", err)
+	}
+
+	contractAddr := tx.To()
+	if *contractAddr != backend.contractAddr {
+		return nil, errors.New("contract address is not supported")
+	}
+
+	// Gas price is not stored in the swap, and is used to determine if the
+	// initialization transaction could take a long time to be mined. A
+	// transaction with a very low gas price may need to be resent with a
+	// higher price.
+	gasPrice, err := ToGwei(tx.GasPrice())
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert gas price: %v", err)
+	}
+
+	// Value is stored in the swap but it is also this value. Because we
+	// can't yet be sure this transaction represents the passed secret
+	// hash, value should be checked by using the contract once it has
+	// confirmations.
+	value, err := ToGwei(tx.Value())
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert value: %v", err)
+	}
+
+	// For redemptions, the transaction should move no value.
+	if sct == sctRedeem && value != 0 {
+		return nil, fmt.Errorf("expected swapCoin value of zero for redeem but got: %d", value)
+	}
+
+	return &swapCoin{
+		backend:      backend,
+		contractAddr: *contractAddr,
+		secret:       secret,
+		secretHash:   secretHash,
+		value:        value,
+		gasPrice:     gasPrice,
+		txid:         hex.EncodeToString(txid[:]),
+		counterParty: *counterParty,
+		locktime:     locktime,
+		sct:          sct,
+	}, nil
+}
+
+// Confirmations returns the number of confirmations for a Coin's
+// transaction. Because a Coin can become invalid after once being
+// considered valid, this condition should be checked for during
+// confirmation counting and an error returned if this Coin is no longer
+// ready to spend. An unmined transaction should have zero confirmations. A
```

Comment:
Does "this condition should be checked" refer to something the caller should do, or something that should be done internal to `Confirmation`?  This sounds like you're describing something `Confirmations` does, and which the caller should recognize.  If so, which error path is it?  The one from `c.backend.node.transaction` (`unable to fetch transaction`)?  Should this be coded?

And is this generally referring to how all eth txns are replacable by making a txn with the same sender nonce, which you are describing in the last paragraph of this method doc, or is this part alluding to something else?